### ISSUE TITLE
`tb_print` and `*_ex` procedures can potentially read past end of buffer

### ISSUE THREAD

**Issue by freddylist** [termbox/termbox2#63]
Hello!

I see that the `utf8_char_to_unicode` does not check whether it has read past a null byte. This allows code like the example below to read past the intended end of a string, at best causing garbage to be printed to the screen or a crash, at worst reading sensitive information. This can happen if we use `tb_print` to print some malcrafted input.

Now the question is am I stupid and sanitizing input is my responsibility before sending it through termbox2? If so, it is in my humble opinion that this should be documented somewhere.

Otherwise, as far as I understand it, the only UTF-8 code with a null byte in it is the null character itself, so termbox2 should not read past null bytes regardless.

### Example:
```c
#define TB_IMPL
#include "termbox2.h"

int main(void) {
	const char ok_string[] = { 'f', 'o', 'o', '\0', 'p', 'a', 's', 's', 'w', 'o', 'r', 'd', '\0' };
	const char bad_string[] = { 'f', 'o', 'o', 194, '\0', 'p', 'a', 's', 's', 'w', 'o', 'r', 'd', '\0' };
	                      /* >2-byte codepoint ^^^    ^ will read past this string terminator! */
	tb_init();
	tb_clear();
	tb_print(0, 0, TB_DEFAULT, TB_DEFAULT, ok_string);
	tb_print(0, 1, TB_DEFAULT, TB_DEFAULT, bad_string);
	tb_print(0, 2, TB_DEFAULT, TB_DEFAULT, "Press ctrl+c to exit.");
	tb_present();
	struct tb_event ev;
	while (tb_poll_event(&ev) == TB_OK) {
		if (ev.key == TB_KEY_CTRL_C)
			break;
	}
	tb_shutdown();
	return 0;
}
```

**Comment by adsr** [comment]
Hello, thanks for filing this issue. I'm surprised fuzz testing never caught this. I agree this should at least be documented and probably warrants some changes.

Right now I'm leaning toward changing `tb_utf8_char_to_unicode` to return an error if it encounters a null char mid sequence, and changing `tb_print*` to replace incomplete chars with `U+FFFD` (replacement character). Any termbox programs hitting this case are already printing garbage (or worse), so I think this is a fine solution. A more "correct" solution might be to let the library user specify a behavior (error, replace, drop, send verbatim, custom callback, etc), but all of that would bloat the API. Programs that want different behavior can sanitize their own input and use `tb_set_cell`.

I'll sleep on it. If anyone has input, please weigh in.

**Comment by pepe** [comment]
The simple solution from the beginning of the second paragraph sounds good. The "correct" solution does not bring value to my library use. The more straightforward API does, tho.

**Comment by rofl0r** [comment]
usual behaviour of C functions is to just return error when invalid input is supplied.
so just print nothing, return -1, or TB_ERR_INVALID_UTF8

**Comment by adsr** [comment]
Fixed in ce6488ccb3789fb7e8ae0d751e3b49bd4e985375. Thank you @freddylist.
